modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java

```diff
@@ -771,6 +772,64 @@ public void handle(HttpExchange exchange) throws IOException {
+        final AtomicInteger numberOfDeleteAttempts = new AtomicInteger(0);
+        final AtomicInteger numberOfSuccessfulDeletes = new AtomicInteger(0);
+        @SuppressForbidden(reason = "use a http server")
+        class ThrottlingDeleteHandler extends S3HttpHandler {
+
+            private final AtomicInteger throttleTimesBeforeSuccess;
+
+            ThrottlingDeleteHandler(int throttleTimesBeforeSuccess) {
+                super("bucket");
+                this.throttleTimesBeforeSuccess = new AtomicInteger(throttleTimesBeforeSuccess);
+            }
+
+            @Override
+            public void handle(HttpExchange exchange) throws IOException {
+                if (exchange.getRequestMethod().equals("POST") && exchange.getRequestURI().toString().startsWith("/bucket/?delete")) {
+                    numberOfDeleteAttempts.incrementAndGet();
+                    if (throttleTimesBeforeSuccess.getAndDecrement() > 0) {
+                        final byte[] responseBody = ("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+                            + "<Error>\n"
+                            + "  <Code>SlowDown</Code>\n"
+                            + "  <Message>This is a throttling message</Message>\n"
+                            + "  <Resource>/bucket/</Resource> \n"
+                            + "  <RequestId>4442587FB7D0A2F9</RequestId>\n"
+                            + "</Error>").getBytes(StandardCharsets.UTF_8);
+
+                        exchange.sendResponseHeaders(HttpStatus.SC_SERVICE_UNAVAILABLE, responseBody.length);
+                        exchange.getResponseBody().write(responseBody);
+                        exchange.getResponseBody().flush();
+                        exchange.close();
+                    } else {
+                        numberOfSuccessfulDeletes.incrementAndGet();
+                        super.handle(exchange);
+                    }
+                } else {
+                    super.handle(exchange);
+                }
+            }
+        }
+
+        List<String> blobsToDelete = new ArrayList<>();
+        for (int i = 0; i < randomIntBetween(500, 3000); i++) {
+            blobsToDelete.add(randomIdentifier());
+        }
+        int throttleTimesBeforeSuccess = randomIntBetween(0, 3);
+        httpServer.createContext("/", new ThrottlingDeleteHandler(throttleTimesBeforeSuccess));
+        blobContainer.deleteBlobsIgnoringIfNotExists(
+            randomFrom(OperationPurpose.SNAPSHOT_DATA, OperationPurpose.SNAPSHOT_METADATA),
+            blobsToDelete.iterator()
+        );
+
+        int expectedNumberOfBatches = (blobsToDelete.size() / 1_000) + (blobsToDelete.size() % 1_000 == 0 ? 0 : 1);
+        assertThat(numberOfDeleteAttempts.get(), equalTo(throttleTimesBeforeSuccess + expectedNumberOfBatches));
+        assertThat(numberOfSuccessfulDeletes.get(), equalTo(expectedNumberOfBatches));
+    }
```

Comment:
This test class seemed very geared towards retrying input stream, but it sounded like the most appropriate home for this.